### ISSUE TITLE
GradientAdjustment L1/L2 regularization issues

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#327]
GradientAdjustment.updateGradientAccordingToParams(...) lines 108 to 111:
(a) At present, L1 and L2 can't/won't be used simultaneously (i.e., if L2, else if L1)
(b) Line 110: test for "conf.getL1() < 0" -> L1 regularization coefficient will/should always be >= 0?
(c) Line 111: Seems significantly wrong. Should be something along the lines of:
```java
	gradient.subi( Transforms.sign(params).muli( conf.getL1()*conf.getLr() ) );
```
though you might want to check that. I'm a little rusty on the mathematics of NNs.

**Comment by AlexDBlack** [comment]
Somewhat related: Momentum implementation (same function as above) seems incorrect.
https://groups.google.com/forum/#!topic/deeplearning4j/EkhV8GhEg30

**Comment by agibsonccc** [comment]
@AlexDBlack  https://github.com/deeplearning4j/deeplearning4j/blob/953b5389324522501484592a5922fa94704af5f9/deeplearning4j-core/src/main/java/org/deeplearning4j/optimize/GradientAdjustment.java#L109-109

Will look at momentum tomorrow. Just pushing up stuff I did on the plane.

**Comment by AlexDBlack** [comment]
Issues (a) and (b) still stand.

As for (c), I'm still not sure that your fix is correct.

I'm not 100% sure what the intent of updateGradientAccordingToParams() is, but re-reading it, my understanding is that given the gradient (array of partial derivatives of loss function wrt parameters (weights? weights/biases?), or gradient from previous layer), we want to modify said gradient vector to account for l1, l2, momentum, adagrad, etc.

(The only other thing I can imagine we are doing here is calculating the parameter update amounts directly, given gradient?)

Referring to the following:
http://neuralnetworksanddeeplearning.com/chap3.html
I think we want to implement as follows: L2 being equation (88) and L1 being equation (96).

If my understanding is correct, I think for L2 the modification (eq. 88) should be:
gradient.addi(params.mul(conf.getL2())
i.e., add w*l2/n, with division by n happening later in method. Present implementation seems to be correct if we are determining the weight update (eq 91).

And for L1 it should be (eq 96)
gradient.addi(Transforms.sign(params).muli(conf.getL1()))
or eq (97) (i.e., eqn in my original issue post) if we are calculating weight updates.

Also, isn't l1/l2 regularization is usually only applied to weights, and not biases? I'm not sure if that is actually what's happening here.

**Comment by agibsonccc** [comment]
You're right about that. I'll add a filter for variable names (basically checks for the variable name bias).

It'd be good to have naming semantics actually mean something anyways.

**Comment by agibsonccc** [comment]
Re: equations. The learning rate is already applied.
